discovery_engine_core/web-api/src/bin/ingestion.rs

```diff
@@ -0,0 +1,349 @@
+}
+
+fn deserialize_string_not_empty_or_zero_byte<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    if s.is_empty() {
+        Err(de::Error::custom("field can't be an empty string"))
+    } else if s.contains('\u{0000}') {
+        Err(de::Error::custom("field can't contain zero bytes"))
+    } else {
+        Ok(s)
+    }
+}
+
+fn deserialize_article_vec_not_empty<'de, D>(deserializer: D) -> Result<Vec<Article>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let v = Vec::deserialize(deserializer)?;
+    if v.is_empty() {
+        Err(de::Error::custom("documents can't be an empty array"))
+    } else {
+        Ok(v)
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), GenericError> {
+    let filter = env::var("RUST_LOG").unwrap_or_else(|_| "tracing=info,warp=debug".to_owned());
+    tracing_subscriber::fmt()
+        .with_env_filter(filter)
+        .with_span_events(FmtSpan::CLOSE)
+        .init();
+
+    let config = Config::init_from_env()?;
+    let client = Client::new();
+    let model = init_model(&config)?;
+
+    let routes = post_add_data(config, model, client)
+        .recover(handle_rejection)
+        .with(warp::trace::request());
+
+    warp::serve(routes).run(([127, 0, 0, 1], 3030)).await;
+
+    Ok(())
+}
+
+fn init_model(config: &Config) -> Result<Model, GenericError> {
+    info!("SMBert model loading...");
+
+    let path = env::current_dir()?;
+    let vocab_path = path.join(&config.smbert_vocab);
+    let model_path = path.join(&config.smbert_model);
+    let smbert = SMBertConfig::from_files(&vocab_path, &model_path)?
+        .with_accents(AccentChars::Cleanse)
+        .with_case(CaseChars::Lower)
+        .with_pooling::<AveragePooler>()
+        .with_token_size(64)?
+        .build()?;
+
+    info!("SMBert model loaded successfully!");
+
+    Ok(Arc::new(smbert))
+}
+
+// POST /add_data
+fn post_add_data(
+    config: Config,
+    model: Model,
+    client: Client,
+) -> impl Filter<Extract = impl Reply, Error = Rejection> + Clone {
+    warp::path("add_data")
+        .and(warp::post())
+        .and(warp::body::content_length_limit(config.max_body_size))
+        .and(warp::body::json())
+        .and(with_model(model))
+        .and(with_config(config))
+        .and(with_client(client))
+        .and_then(handle_add_data)
+}
+
+async fn handle_add_data(
+    body: AddDataRequestBody,
+    model: Model,
+    config: Config,
+    client: Client,
+) -> Result<impl warp::Reply, Rejection> {
+    if body.documents.len() > config.max_documents_length {
+        return Err(warp::reject::custom(TooManyDocumentsError));
+    }
+
+    let (documents, errored_ids): (Vec<_>, Vec<_>) = body
+        .documents
+        .into_iter()
+        .map(|mut article| match model.run(&article.snippet) {
+            Ok(embedding) => {
+                article.embedding = embedding.iter().cloned().collect();
```

Comment:
:grey_question: Isn't `embedding.to_vec()` enough?

Reply:
We can also do
```rust
if embedding.is_standard_layout() { embedding.into_raw_vec() } else { embedding.to_vec() }
```

Reply:
`to_vec` was enough, the other suggestion worked when I've adjusted it to:
```
embedding.as_standard_layout().to_owned().into_raw_vec()
```